incubator/virtualcluster/pkg/syncer/fairqueue/fair.go

```diff
@@ -0,0 +1,226 @@
+	// element of queue should be in the dirty set and not in the
+	// processing set.
+	queueGroup map[string][]t
+
+	// dirty defines all of the items that need to be processed.
+	dirty set
+
+	// Things that are currently being processed are in the processing set.
+	// These things may be simultaneously in the dirty set. When we finish
+	// processing something and remove it from this set, we'll check if
+	// it's in the dirty set, and if so, add it to the queue.
+	processing set
+
+	cond *sync.Cond
+
+	shuttingDown bool
+
+	rateLimiter workqueue.RateLimiter
+
+	// clock tracks time for delayed firing
+	clock clock.Clock
+
+	// stopCh lets us signal a shutdown to the waiting loop
+	stopCh chan struct{}
+
+	// heartbeat ensures we wait no more than maxWait before firing
+	heartbeat clock.Ticker
+
+	// waitingForAddCh is a buffered channel that feeds waitingForAdd
+	waitingForAddCh chan *waitFor
+}
+
+type empty struct{}
+type t interface{}
+type set map[t]empty
+
+func (s set) has(item t) bool {
+	_, exists := s[item]
+	return exists
+}
+
+func (s set) insert(item t) {
+	s[item] = empty{}
+}
+
+func (s set) delete(item t) {
+	delete(s, item)
+}
+
+func NewRateLimitingFairQueue(rateLimiter workqueue.RateLimiter) workqueue.RateLimitingInterface {
+	return newRateLimitingFairQueue(clock.RealClock{}, rateLimiter)
+}
+
+func newRateLimitingFairQueue(clock clock.RealClock, rateLimiter workqueue.RateLimiter) workqueue.RateLimitingInterface {
+	ret := &fairQueue{
+		balancer:        wrr.NewWeightedRR(),
+		queueGroup:      make(map[string][]t),
+		dirty:           set{},
+		processing:      set{},
+		rateLimiter:     rateLimiter,
+		cond:            sync.NewCond(&sync.Mutex{}),
+		clock:           clock,
+		heartbeat:       clock.NewTicker(maxWait),
+		stopCh:          make(chan struct{}),
+		waitingForAddCh: make(chan *waitFor, 1000),
+	}
+
+	go ret.waitingLoop()
+
+	return ret
+}
+
+func (q *fairQueue) Add(obj interface{}) {
+	q.cond.L.Lock()
+	defer q.cond.L.Unlock()
+	if q.shuttingDown {
+		return
+	}
+	item, ok := obj.(GroupItem)
+	if !ok {
+		return
+	}
+
+	if q.dirty.has(item) {
+		return
+	}
+
+	q.dirty.insert(item)
+	if q.processing.has(item) {
+		return
+	}
+
+	group := item.Group()
+	_, exists := q.queueGroup[group]
+	if !exists {
+		q.queueGroup[group] = []t{}
+		q.balancer.Add(group, 1)
+	}
+
+	q.queueGroup[group] = append(q.queueGroup[group], item)
+	q.cond.Signal()
+}
+
+func (q *fairQueue) Len() int {
+	q.cond.L.Lock()
+	defer q.cond.L.Unlock()
+	return q.len()
+}
+
+func (q *fairQueue) len() int {
+	var total = 0
+	for _, q := range q.queueGroup {
+		total += len(q)
+	}
+	return total
+}
+
+func (q *fairQueue) Get() (item interface{}, shutdown bool) {
+	q.cond.L.Lock()
+	defer q.cond.L.Unlock()
+	for q.len() == 0 && !q.shuttingDown {
+		q.cond.Wait()
+	}
+	if q.len() == 0 {
+		// We must be shutting down.
+		return nil, true
+	}
+
+	var nextGroup string
+	// TODO(zhuangqh): asynchronously queue gc.
+	for {
```

Comment:
This can be inefficient if there are many tenants but only a few are actively (common in VC use cases). A lot of time will be spent in examining empty queues. 

We should keep a list of indexes for actively queues and apply wrr on active queues only Or we can gc queue when queue length becomes zero. 

This should be addressed in next change.